src/west/commands/project.py

```diff
@@ -817,12 +855,43 @@ def _update_special(name):
                           '(url)).')
 
             if project.name == 'west':
-                # Signal self-update, which will cause a restart. This is a
-                # bit nicer than doing the restart here, as callers will
-                # have a chance to flush file buffers, etc.
+                # Signal self-update, which will cause a restart. This is a bit
+                # nicer than doing the restart here, as callers will have a
+                # chance to flush file buffers, etc.
                 raise WestUpdated()
 
 
+def _update_and_reset_special(name):
+    # Updates one of the special repositories (the manifest and west) by
+    # resetting to the new revision after fetching it (with 'git reset --keep')
+
+    project = _special_project(name)
+    with _error_context(', while updating/resetting special project'):
+        _inf(project,
+             "Fetching and resetting (name-and-path) to '(revision)'")
+        _git(project, 'fetch -- (url) (revision)')
+        if _git(project, 'reset --keep FETCH_HEAD', check=False).returncode:
```

Comment:
Combined with the way we a repo is cloned in _bootstrap: `git clone -b <rev> -- <url> <dest>`,
resetting this way, will result the repo to still be refer to the original checked out branch, however, now containing the FETCH_HEAD code.
This causes the the local branch to differ from upstream, see below.

I suggest that when bootstrap inits (clones) the remote repo, then we create and switch to local special branch, i.e. similar to what is done for the project, where `manifest-rev` is local only branch, and we go into detached head.

Example, seen from user:
```
$ west init --mr work_branch_1  <-- starting with work_branch_1
=== Updated configuration written to /projects/github/west_test/west/config ===                        
=== Running 'west update --reset-manifest --reset-projects' to update repositories ===                 
=== Fetching and resetting manifest (west/manifest/) to 'work_branch_1'                                
                                                                                                       
$ pushd west/manifest/                                                                                 
$ git status                                                                                           
On branch work_branch_1   <-- We are at local branch work_branch_1, which tracks origin/work_branch_1
Your branch is up-to-date with 'origin/work_branch_1'.                                                 
nothing to commit, working directory clean                                                             
$ popd                                                                                                 
                                                                                                       
$ west init --mr work_branch_2                                                                         
=== Updated configuration written to /projects/github/west_test/west/config ===                        
=== Running 'west update --reset-manifest --reset-projects' to update repositories ===                 
=== Fetching and resetting manifest (west/manifest/) to 'work_branch_2'                                
                                                                                                       
$ pushd west/manifest/                                                                                 
$ git status                                                                                           
On branch work_branch_1 <-- We are still at work_branch_1, which tracks origin/work_branch_1, but now with content of work_branch_2, which could be very different content.                                                                               
Your branch and 'origin/work_branch_1' have diverged,                                                  
and have 1 and 1 different commit each, respectively.                                                  
(use "git pull" to merge the remote branch into yours)                                                 
nothing to commit, working directory clean                                                             
$ popd                                                                                                 
```

Reply:
Think it's similar to the things I suggested in https://github.com/zephyrproject-rtos/west/pull/99#discussion_r233570898. Another limitation at the moment is that you can only pass a branch name to `--mr <rev>` when doing the initial clone, because it turns into `git clone -b <rev>`.

What do you mean by this part? Should it both create a `manifest-rev` and check out a detached HEAD?

```
I suggest that when bootstrap inits (clones) the remote repo, then we create and switch to local special branch, i.e. similar to what is done for the project, where manifest-rev is local only branch, and we go into detached head.
```

Reply:
FWIW, I'm not a big fan of the project resetting behavior either, because it will mess with whatever branches people have checked out in the same way. I'm just following specs though.

Reply:
> 
> What do you mean by this part? Should it both create a `manifest-rev` and check out a detached HEAD?
> 

I was simply referring to how it is done for the projects themselves, e.g. where we end in detached head.
So consistency, and as well to avoid placing changes on a tracking branchm where they do not belong (unless user does so on purpose).

Btw. I have to correct myself. The detached head doesn't happen on the `clone`, but later on the `pull` stage, see below.
Having same behavior for special repos (west/manifest) might make sense, unless there are special reasons not to do so.

Command example
```
$ west clone  <-- Only clones the repo and check out master branch, although it does create the manifest-rev branch
=== Creating repository for zephyr (zephyr/)                                                        
Initialized empty Git repository in /projects/github/west_test/zephyr/.git/                         
=== Fetching changes for zephyr (zephyr/)                                                           
.....                                                                                               
Checking out files: 100% (11967/11967), done.                                                       
Note: checking out 'refs/heads/manifest-rev'.                                                       
                                                                                                    
......                                                                                              
=== Creating branch 'master' in zephyr (zephyr/)                                                    
=== Checking out branch 'master' in zephyr (zephyr/)                                                
Switched to branch 'master'                                                                         
Your branch is up-to-date with 'manifest-rev'.                                                      
$ cd zephyr/                                                                                        
$ git branch                                                                                        
manifest-rev                                                                                        
* master                                                                                            
$ cd ..                                                                                             
$ west pull  <-- This gets us in detached head state                                                                                       
=== Fetching changes for zephyr (zephyr/)                                                           
HEAD is now at d0253d3... samples: reel_board/mesh_badge: Fix various coding style issues           
=== zephyr (zephyr/) is up-to-date with manifest-rev                                                
$ cd zephyr/                                                                                        
$ git branch                                                                                        
* (HEAD detached at refs/heads/manifest-rev) 
manifest-rev                                                                                        
master                     
```

Reply:
> Think it's similar to the things I suggested in [#99 (comment)](https://github.com/zephyrproject-rtos/west/pull/99#discussion_r233570898). Another limitation at the moment is that you can only pass a branch name to `--mr <rev>` when doing the initial clone, because it turns into `git clone -b <rev>`.
> 
I didn't read your the comment a similar suggestion, but it's nice if were are aligned ;)

I agree that it is a limitation in the existing `west init` behavior, caused by the `git clone -b`

So maybe this should be put as an improvement on the bootstapper ?
but before doing that, we should ensure that we then obtain the desired behavior with this code.
As well as comments from the rest.

Reply:
@ulfalizer Created: #110 as I don't think this issue should block the PR, as it  is more an edge case for time being, and in order for us to not forget.